crates/services/txpool_v2/src/storage/graph.rs

```diff
@@ -0,0 +1,555 @@
+use std::{
+    collections::{
+        HashMap,
+        HashSet,
+    },
+    time::Instant,
+};
+
+use fuel_core_types::{
+    fuel_tx::{
+        input::{
+            coin::{
+                CoinPredicate,
+                CoinSigned,
+            },
+            contract::Contract,
+            message::{
+                MessageCoinPredicate,
+                MessageCoinSigned,
+                MessageDataPredicate,
+                MessageDataSigned,
+            },
+        },
+        ContractId,
+        Input,
+        Output,
+        TxId,
+        UtxoId,
+    },
+    services::txpool::PoolTransaction,
+};
+use petgraph::{
+    graph::NodeIndex,
+    prelude::StableDiGraph,
+};
+
+use crate::{
+    collision_manager::{
+        basic::BasicCollisionManagerStorage,
+        CollisionReason,
+    },
+    error::Error,
+    ports::TxPoolPersistentStorage,
+    selection_algorithms::ratio_tip_gas::RatioTipGasSelectionAlgorithmStorage,
+};
+
+use super::{
+    RemovedTransactions,
+    Storage,
+    StorageData,
+};
+
+pub struct GraphStorage {
+    /// The configuration of the graph
+    config: GraphConfig,
+    /// The graph of transactions
+    graph: StableDiGraph<StorageData, ()>,
+    /// Coins -> Transaction that currently create the UTXO
+    coins_creators: HashMap<UtxoId, NodeIndex>,
+    /// Contract -> Transaction that currently create the contract
+    contracts_creators: HashMap<ContractId, NodeIndex>,
+}
+
+pub struct GraphConfig {
+    /// The maximum number of transactions per dependency chain
+    pub max_txs_chain_count: usize,
+}
+
+impl GraphStorage {
+    /// Create a new graph storage
+    pub fn new(config: GraphConfig) -> Self {
+        Self {
+            config,
+            graph: StableDiGraph::new(),
+            coins_creators: HashMap::new(),
+            contracts_creators: HashMap::new(),
+        }
+    }
+}
+
+impl GraphStorage {
+    /// Remove a node and all its dependent sub-graph.
+    /// Edit the data of dependencies transactions accordingly.
+    /// Returns the removed transactions.
+    fn remove_node_and_dependent_sub_graph(
+        &mut self,
+        root_id: NodeIndex,
+    ) -> Result<Vec<PoolTransaction>, Error> {
+        let Some(root) = self.graph.node_weight(root_id) else {
+            return Ok(vec![])
+        };
+        let gas_removed = root.dependents_cumulative_gas;
+        let tip_removed = root.dependents_cumulative_tip;
+        let dependencies: Vec<NodeIndex> = self.get_dependencies(root_id)?.collect();
+        let removed = self.remove_dependent_sub_graph(root_id)?;
```

Comment:
I think we need to call `self.reduce_dependencies_cumulative_gas_tip_and_chain_count` for all children in the subgraph as well. Since they can have several parents(and some of these parents are not parents of the `root_id`).

Reply:
Done